### PR TITLE
Refactor Protobuf generation to use Buf dependencies

### PR DESCRIPTION
**Pull request by austinchandra**
## Changes
See [review doc](https://www.notion.so/evmos/EvmosJS-Protobuf-Refactor-Documentation-b9d461289ee44fe1ae8497f9ac2bd82c) for context and coverage of all changes.

By executing this refactor, we remove hundreds of files from our codebase, add necessary testing infrastructure, and organize the project around a logical architecture.

Despite its size, this PR is designed to be reviewable (or so I hope). The vast majority of changed files are deleted files, single-line re-exports, or moved files—the remaining files are unit tests.

### CODE REVIEW

File: packages/proto/src/transaction/transaction.spec.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
+import { createMsgSend } from '../messages/bank/msgSend'
```
Comment (austinchandra) [inline packages/proto/src/transaction/transaction.spec.ts:1]: Renaming of `packages/proto/tests/transaction/transaction.spec.ts`

File: packages/proto/tests/transaction/transaction.spec.ts [hunk 2]
```diff
@@ -1,228 +0,0 @@
...
-      37, 92, 174, 75, 160, 196, 188, 55, 202, 114, 97, 5, 178, 20, 10, 253, 14,
-      105, 23,
-    ])
-    const sequence = 0
-    const info = createSignerInfo('ethsecp256k1', pubkey, sequence, SIGN_DIRECT)
-    expect(info.toObject()).toStrictEqual({
-      public_key: {
-        type_url: '/ethermint.crypto.v1.ethsecp256k1.PubKey',
-        // value: 'CiMKIQKIsfUxuHhx28A3KVGHJVyuS6DEvDfKcmEFshQK/Q5pFw=='
-        value: new Uint8Array([
```
Comment (austinchandra) [inline packages/proto/tests/transaction/transaction.spec.ts:72]: @hanchon Do you know why this `public_key` value doesn't match the original bytes?

After refactoring the Protobuf generation library, we compare it directly to the `pubkey` value, which passes; will this cause any downstream issues?

File: packages/proto/src/messages/distribution/distribution.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+export interface MsgWithdrawDelegatorRewardProtoInterface {
+  path: string
+  message: MsgWithdrawDelegatorReward
+}
+
+export function createMsgWithdrawDelegatorReward(
+  delegatorAddress: string,
+  validatorAddress: string,
+) {
```
Comment (fedekunze) [inline packages/proto/src/messages/distribution/distribution.ts:15]: this file is not very readable because of all the `exports`, can we group them?

File: packages/proto/src/messages/erc20/erc20.spec.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { createMsgConvertCoin } from './msgConvertCoin'
+import { createMsgConvertERC20 } from './msgConvertERC20'
+
+import { MsgConvertCoin, MsgConvertERC20 } from '../../proto/evmos/erc20/tx'
+
+import { from, to, denom, hex } from '../../proto/tests/utils'
+import { JSONOptions } from '../../proto/tests/common'
+
+describe('test erc20 message generation', () => {
+  it('msgConvertCoin', () => {
```
Comment (fedekunze) [inline packages/proto/src/messages/erc20/erc20.spec.ts:10]: nit, test naming is not clear

File: packages/proto/src/messages/evm/web3Extension.spec.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import { createWeb3Extension } from './web3Extension'
+
+import { ExtensionOptionsWeb3Tx } from '../../proto/ethermint/types/web3'
+
+import { from } from '../../proto/tests/utils'
+import { JSONOptions } from '../../proto/tests/common'
+
+describe('test web3Extension message generation', () => {
+  it('web3Extension', () => {
```
Comment (fedekunze) [inline packages/proto/src/messages/evm/web3Extension.spec.ts:9]: are we only testing success cases?

File: packages/transactions/src/messages/web3Extension.spec.ts [hunk 6]
```diff
@@ -19,7 +19,7 @@ describe('web3Extension tests', () => {
 
     const res = signatureToWeb3Extension(chain, sender, signature)
     expect(res.path).toBe('ethermint.types.v1.ExtensionOptionsWeb3Tx')
-    expect(Buffer.from(res.message.serializeBinary()).toString('base64')).toBe(
+    expect(Buffer.from(res.message.toBinary()).toString('base64')).toBe(
```
Comment (fedekunze) [inline packages/transactions/src/messages/web3Extension.spec.ts:22]: why the change from `serializeBinary()`, to `toBinary()`?

**Review by fedekunze — COMMENTED** [review]
Suggestions:

- Group Exports of functions and interfaces.
- test error cases on message generation. Looks like we're only testing for success
- nit, tests case names are not very clear
- nit, add comments on exports

File: packages/transactions/src/messages/web3Extension.spec.ts (hunk 6 above)
Comment (austinchandra) [reply packages/transactions/src/messages/web3Extension.spec.ts:22]: We switched Protobuf Typescript packages, the new one has renamed methods and various changes, one of which is this method

File: packages/proto/src/messages/evm/web3Extension.spec.ts (hunk 5 above)
Comment (austinchandra) [reply packages/proto/src/messages/evm/web3Extension.spec.ts:9]: Really, there are no failure cases for these methods. The tests ensure that each parameter gets handled as expected, and we don't accidentally do anything stupid, like exclude a field passed in.

File: packages/proto/src/messages/crypto/crypto.spec.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import { createED25519PubKey } from './keys'
+import { PubKey } from '../../proto/cosmos/crypto/ed25519/keys'
+import { JSONOptions } from '../../proto/tests/common'
+
+describe('test crypto message generation', () => {
+  it('createED25519PubKey', () => {
```
Comment (fedekunze) [inline packages/proto/src/messages/crypto/crypto.spec.ts:6]: ditto

File: packages/proto/src/messages/evm/msgEthereumTx.spec.ts [hunk 8]
```diff
@@ -30,41 +31,40 @@ describe('msgEthereumTx tests', () => {
...
     expect(bodyProtoMessages[0].value).toBeDefined()
     const msgEthereumTxProto = bytesToMsgEthereumTx(
       bodyProtoMessages[0].value as Uint8Array,
     )
-    const msgEthTx = msgEthereumTxProto.toObject()
+    const msgEthTx = msgEthereumTxProto
 
     // Create the LegacyTx/AccessListTx/DynamicFeeTx depending on the type_url
-    expect(msgEthTx.data?.type_url).toBe('/ethermint.evm.v1.LegacyTx')
+    expect(msgEthTx.data?.typeUrl).toBe(`/${LegacyTx.typeName}`)
```
Comment (fedekunze) [inline packages/proto/src/messages/evm/msgEthereumTx.spec.ts:56]: Let's also test the other data options

File: packages/proto/src/messages/ibc/ibcMsgTransfer.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import { Coin } from '../../proto/cosmos/base/coin'
+import { Height } from '../../proto/cosmos-ibc/ibc/core/client'
+import { MsgTransfer } from '../../proto/cosmos-ibc/ibc/applications/tx'
+
+export function createIBCMsgTransfer(
```
Comment (fedekunze) [inline packages/proto/src/messages/ibc/ibcMsgTransfer.ts:5]: memo is missing

File: packages/proto/src/messages/ibc/ibc.spec.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        denom: params.denom,
+      },
+      sender: params.sender,
+      receiver: params.receiver,
+      timeout_height: {
+        revision_number: params.revisionNumber.toString(),
+        revision_height: params.revisionHeight.toString(),
+      },
+      timeout_timestamp: params.timeoutTimestamp.toString(),
+      memo: '',
```
Comment (fedekunze) [inline packages/proto/src/messages/ibc/ibc.spec.ts:48]: `memo` should be included in the constructor args

File: packages/proto/src/messages/index.ts [hunk 11]
```diff
@@ -0,0 +1,15 @@
+export * from './authz'
+export * from './bank'
+export * from './crypto'
+export * from './distribution'
+export * from './erc20'
+export * from './evm'
+export * from './gov'
+export * from './ibc'
+export * from './revenue'
+// TODO: Find out why export * from './staking' doesn't work
```
Comment (fedekunze) [inline packages/proto/src/messages/index.ts:10]: nit, create ticket for this

File: packages/proto/src/messages/staking/staking.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
+import { Coin } from '../../proto/cosmos/base/coin'
+import {
+  MsgDelegate,
+  MsgBeginRedelegate,
+  MsgUndelegate,
+} from '../../proto/cosmos/staking/tx'
```
Comment (fedekunze) [inline packages/proto/src/messages/staking/staking.ts:6]: cancel unbond is missing

File: packages/proto/src/messages/vesting/vesting.spec.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
+import { createMsgClawback } from './msgClawback'
+import { createMsgCreateClawbackVestingAccount } from './msgCreateClawbackVestingAccount'
+
+import {
+  MsgCreateClawbackVestingAccount,
+  MsgClawback,
+} from '../../proto/evmos/vesting/tx'
```
Comment (fedekunze) [inline packages/proto/src/messages/vesting/vesting.spec.ts:7]: the new txs types are missing

**Review by fedekunze — CHANGES_REQUESTED** [review]
Thanks @austinchandra, minor comments

- A few tx types are missing on staking and vesting
- The `memo` field is missing on IBC transfers
- Please also update the test names for better readability
- Add tests for the other Ethereum Tx data types

File: .github/workflows/test.yml [hunk 14]
```diff
@@ -23,13 +23,13 @@ jobs:
             yarn.lock
 
       - name: Install modules
-        run: yarn install
+        run: npm install
```
[the remark this answers is not in the document]
Comment (austinchandra) [reply .github/workflows/test.yml:26]: Actually, this change is related. `buf`'s NPM registry does not work with Yarn under `v2`, and updating Yarn is a breaking change. In order to build the project, we need to switch to `npm`.

File: packages/proto/src/messages/ibc/ibcMsgTransfer.ts (hunk 9 above)
Comment (austinchandra) [reply packages/proto/src/messages/ibc/ibcMsgTransfer.ts:5]: We should create a separate PR for this, since we would need to add it to `transactions` and `eip712` as well.

File: packages/proto/src/messages/staking/staking.ts (hunk 12 above)
Comment (austinchandra) [reply packages/proto/src/messages/staking/staking.ts:6]: We should separately add this to all packages in a separate PR.

File: packages/proto/src/messages/vesting/vesting.spec.ts (hunk 13 above)
Comment (austinchandra) [reply packages/proto/src/messages/vesting/vesting.spec.ts:7]: Same as above

File: packages/proto/src/messages/distribution/distribution.ts (hunk 3 above)
Comment (austinchandra) [reply packages/proto/src/messages/distribution/distribution.ts:15]: @hanchon What exactly are `MsgWithdrawValidatorCommissionProtoInterface` and `MsgSetWithdrawAddressProtoInterface` used for? They're not imported elsewhere in the codebase. Can we delete them?

File: packages/proto/src/messages/ibc/ibcMsgTransfer.ts (hunk 9 above)
Comment (austinchandra) [reply packages/proto/src/messages/ibc/ibcMsgTransfer.ts:5]: I want to minimize the scope of this PR as much as possible, so we can merge it in and unblock other work. I can open PRs for these fixes as soon as we finalize this refactor.

File: packages/proto/src/messages/distribution/distribution.ts (hunk 3 above)
Comment (hanchon) [reply packages/proto/src/messages/distribution/distribution.ts:15]: No, the functions are being used by evmos.me to generate that transaction, @fedekunze asked for that transaction some time ago so I just added the helpers here to make a quick release

File: packages/proto/tests/transaction/transaction.spec.ts (hunk 2 above)
Comment (hanchon) [reply packages/proto/tests/transaction/transaction.spec.ts:72]: #134 with that patch applied I was able to send transactions with metamask and keplr

File: packages/proto/src/messages/ibc/ibcMsgTransfer.ts (hunk 9 above)
Comment (fedekunze) [reply packages/proto/src/messages/ibc/ibcMsgTransfer.ts:5]: They are different. The memo I'm referring to goes in the IBC packet, not in the transaction.

File: packages/proto/src/messages/staking/staking.ts (hunk 12 above)
Comment (fedekunze) [reply packages/proto/src/messages/staking/staking.ts:6]: why in a separate PR? doesn't this mean that the staking proto is outdated? This should be addressed in this PR

File: packages/proto/src/messages/vesting/vesting.spec.ts (hunk 13 above)
Comment (fedekunze) [reply packages/proto/src/messages/vesting/vesting.spec.ts:7]: see reply

File: packages/proto/src/messages/ibc/ibcMsgTransfer.ts (hunk 9 above)
Comment (austinchandra) [reply packages/proto/src/messages/ibc/ibcMsgTransfer.ts:5]: I understand, but this would break the code within the `transactions` package that calls the `proto` implementation. Furthermore, the `transactions` package wouldn't support receiving the IBC `memo` field as a parameter until we add it, and the `EIP-712` package would not support it either.

In order to fully support this change, we have to make changes across `transactions` and `eip712`. That should be done separately.

File: packages/proto/src/messages/staking/staking.ts (hunk 12 above)
Comment (austinchandra) [reply packages/proto/src/messages/staking/staking.ts:6]: The staking Proto is outdated, as is the rest of the codebase. The point of this PR is to refactor the `proto` code so that it's maintainable.

When it comes to messages, we can add support across the entire codebase at once.

**Review by hanchon — APPROVED** [review]

File: packages/proto/src/messages/ibc/ibcMsgTransfer.ts (hunk 9 above)
Comment (fedekunze) [reply packages/proto/src/messages/ibc/ibcMsgTransfer.ts:5]: ok, your plan sounds good to me. Can you create a ticket for it to add it to v12?

File: packages/proto/src/messages/staking/staking.ts (hunk 12 above)
Comment (fedekunze) [reply packages/proto/src/messages/staking/staking.ts:6]: sg, please create a ticket to track the task

**Review by fedekunze — APPROVED** [review]
ACK, could you create tickets for the following outstanding comments?

- [ ] add missing txs (vesting, staking)
- [ ] support memo field for IBC transfer
- [ ] add tests cases for other Ethereum tx data types (dynamic fee tx, access list tx, etc)

**Comment by austinchandra** [comment]
Added tickets as sub-tasks within the [main issue](https://linear.app/evmos/issue/ENG-1535/evmosjs-modernization).
